### PR TITLE
Ensure InputTracker tracks changing mount paths

### PR DESCRIPTION
**Pull request by anniefu**
When checking whether the build graph can be re-used, InputTracker should take into account any changes to build mount paths and automatically return a graph cache miss if there are any changes. 

Otherwise, a graph might be re-used with the wrong mount paths causing incorrect tracking and possibly Disallowed File Accesses for the new mount paths. For example, when different users share the same EngineCache and have different user profile mount paths.

### CODE REVIEW

File: Public/Src/Engine/Dll/InputTracker.cs [hunk 1]
```diff
@@ -530,7 +530,9 @@ private void RegisterFile(string path, ContentHash hash)
         /// </summary>
         public void WriteToFile(
             BinaryWriter writer,
+            PathTable pathTable,
             IReadOnlyDictionary<string, string> buildParametersImpactingBuild,
+            IReadOnlyDictionary<string, IMount> mountsImpactingBuild,
```
Comment (dannyvv) [inline Public/Src/Engine/Dll/InputTracker.cs]: Nit:perhaps annotate with [CanBeNull] so we get IDE help when calling this function.

File: Public/Src/Engine/Dll/InputTracker.cs [hunk 2]
```diff
@@ -570,6 +572,23 @@ private void RegisterFile(string path, ContentHash hash)
                 }
             }
 
+            // Write the mounts that impact the build
+            if (mountsImpactingBuild == null)
+            {
+                writer.Write(0);
```
Comment (dannyvv) [inline Public/Src/Engine/Dll/InputTracker.cs]: tiny nitpick: Upon deserialization we can't distinguish empty mounts from null mounts...

File: Public/Src/Engine/Dll/Strings.resx [hunk 3]
```diff
@@ -216,4 +216,10 @@
   <data name="Layout_DefaultJunctionNameTotLogFolder" xml:space="preserve">
     <value>BuildXLCurrentLog</value>
   </data>
+  <data name="InputTracker_MountChanged" xml:space="preserve">
+    <value>Mount {0} was:'{1}' now:'{2}'</value>
+  </data>
+  <data name="InputTracker_MountRemoved" xml:space="preserve">
+    <value>{0} was previously defined as a mount but is now undefined</value>
```
Comment (dannyvv) [inline Public/Src/Engine/Dll/Strings.resx]: would it be usefull to report what path it was previously defined with?

File: Public/Src/Engine/Dll/InputTracker.cs (hunk 2 above)
Comment (anniefu) [reply Public/Src/Engine/Dll/InputTracker.cs]: I don't think that's that big of a deal. But are you allowed to have a mount that has no name in the first place?

Comment (anniefu) [reply Public/Src/Engine/Dll/InputTracker.cs]: I take that back, empty mount names don't make a difference here actually.

I'm just going to leave it as-is though to match the pattern that the environment variables dictionary above uses for serialization.

**Comment by narasamdya** [comment]
>             if ((missReasonForCompatibleGraphFingerprint = previousFingerprint.CompareFingerprint(graphFingerprint.CompatibleFingerprint)) == GraphCacheMissReason.NoMiss)

Can we bump up the graph fingerprint version so that the new BuildXL will not throw exception (if possible) when trying to load old engine cache?

---
Refers to: Public/Src/Engine/Dll/InputTracker.cs:826 in 1021aee. [](commit_id = 1021aee7eb3f741acf49f9b8d6c58958815fad05, deletion_comment = False)

File: Public/Src/Engine/UnitTests/Engine/MiniBuild.cs [hunk 4]
```diff
@@ -650,6 +650,7 @@ public void MiniBuildCachedGraphWithMountAndEnvironmentChanges()
             // Revert environment variable.
             Configuration.Startup.Properties[EnvVarName] = "Env1";
 
+            System.Diagnostics.Debugger.Launch();
```
Comment (narasamdya) [inline Public/Src/Engine/UnitTests/Engine/MiniBuild.cs]: >System.Diagnostics.Debugger.Launch(); [](start = 12, length = 37)

Remove :-)

File: Public/Src/Engine/Dll/InputTracker.cs (hunk 2 above)
Comment (dannyvv) [reply Public/Src/Engine/Dll/InputTracker.cs]: sounds good, was merely observigin the difference

---
In reply to: [274223325](https://github.com/Microsoft/BuildXL/pull/126#discussion_r274223325) [](ancestors = 274223325)

**Pull request merged**